pkgs/by-name/wi/wivrn/package.nix

```diff
@@ -0,0 +1,120 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, fetchFromGitLab
+, avahi
+, boost
+, cmake
+, cudaPackages
+, eigen
+, ffmpeg
+, freetype
+, git
+, glm
+, glslang
+, harfbuzz
+, libdrm
+, libva
+, libpulseaudio
+, libX11
+, libXrandr
+, nlohmann_json
+, onnxruntime
+, openxr-loader
+, pkg-config
+, python3
+, shaderc
+, spdlog
+, systemd
+, udev
+, vulkan-headers
+, vulkan-loader
+, vulkan-tools
+, x264
+}:
+stdenv.mkDerivation (finalAttrs: {
+  pname = "wivrn";
+  version = "0.12";
+
+  src = fetchFromGitHub {
+    owner = "meumeu";
+    repo = "wivrn";
+    rev = "v${finalAttrs.version}";
+    hash = "sha256-O6Eq7EQ427hOcN16Z33I74CevnHlX/a4ZAcljgc+vk8=";
+  };
+
+  monadoSrc = fetchFromGitLab {
+    domain = "gitlab.freedesktop.org";
+    owner = "monado";
+    repo = "monado";
+    # Version stated in CMakeList for WiVRn 0.12
+    rev = "ffb71af26f8349952f5f820c268ee4774613e200";
+    hash = "sha256-+RTHS9ShicuzhiAVAXf38V6k4SVr+Bc2xUjpRWZoB0c=";
+  };
+
+  # The library path to the OpenXR runtime requires a relative path from the config file to the binary in the nix store
+  # The config file is usually located at ~/.config/openxr/1/ but the wivrn module puts it at /etc/xdg/openxr/1/
+  # The CMakeList has relative directory paths that cause malformation of the path. https://github.com/Meumeu/WiVRn/issues/47
+  # What it is: ../../..//nix/store/...
+  # What it should be: ../../../../nix/store/...
+  # Details about the required path here (Section 3): https://monado.freedesktop.org/valve-index-setup.html
+  patchPhase = ''
+    substituteInPlace ./server/CMakeLists.txt \
+      --replace "../../../" "../../../.."
+  '';
```

Comment:
Does it really *require* a relative path? I wonder if you could just replace `../../../` with am empty string.

Reply:
I have tested it without relative paths and it does appear to work but I'm not sure if that is intended functionality. I've simply just gone by what Monado/OpenComposite says to do.

Reply:
I am not quite sure here. The existing Nix packages for Monado generate a manifest with an absolute path.

According to the [OpenXR Loader spec](https://registry.khronos.org/OpenXR/specs/1.0/loader.html#runtime-manifest-file-format), both absolute and relative paths are supported. I think an absolute path is definitely preferable here.

Reply:
Oh that's real nice then haha
I think all that's left to figure out is where to generate the `active_runtime.json` so it can be found by Steam

Reply:
I think we should attach a link to https://github.com/NixOS/nixpkgs/pull/293058#discussion_r1524376591 in the comment as well.

Reply:
>   # What it is: ../../..//nix/store/...
>  # What it should be: ../../../../nix/store/...

This can be summarised as abusing `${prefix}/CMAKE_INSTALL_${COMPONENT}DIR` instead of `${CMAKE_INSTALL_FULL_${COMPONENT}DIR`

Reply:
> Oh that's real nice then haha I think all that's left to figure out is where to generate the `active_runtime.json` so it can be found by Steam

That's not really in the scope of this package. What do you mean *by steam*? It's the OXR loader's job?

Reply:
> > `# What it is: ../../..//nix/store/...`
> > `# What it should be: ../../../../nix/store/...`
> 
> This can be summarised as abusing `${prefix}/CMAKE_INSTALL_${COMPONENT}DIR` instead of `${CMAKE_INSTALL_FULL_${COMPONENT}DIR`

I changed that comment, look at the updated one

> That's not really in the scope of this package. What do you mean _by steam_? It's the OXR loader's job?

More so just figuring out where to place it so that it gets found in a system-wide path. It should be found in `/etc/xdg/openxr/1/active_runtime.json` as described in [here](https://github.com/KhronosGroup/OpenXR-SDK-Source/blob/main/specification/loader/runtime.adoc) but I am unsuccessful in that regard.

Reply:
> More so just figuring out where to place it so that it gets found in a system-wide path. It should be found in `/etc/xdg/openxr/1/active_runtime.json` as described in [here](https://github.com/KhronosGroup/OpenXR-SDK-Source/blob/main/specification/loader/runtime.adoc?rgh-link-date=2024-03-17T18%3A17%3A04Z) but I am unsuccessful in that regard.

That's the job of the NixOS module and it's already being handled there.

Reply:
Is it a steam game? `/etc` is being overridden by a container. Home dir is passed through, and WiVRn sets the active_runtime there once a client is connected and Monado is started.

Reply:
Have you tried xrgears? The path here should definitely be made absolute since it's set in both ~/.config and /etc/xdg and those are different depths. [My personal branch](https://github.com/ckiee/nixpkgs/tree/c21c0114f7dec00d414b2f6d47839660b786c96d/pkgs/by-name/wi/wivrn/package.nix) has `--replace "../../../" "../../../../../.."` and that makes it work for `~/.config`, so this fails:

```
[nix-shell:~]$ rm ~/.config/openxr/1/active_runtime.json
 
[nix-shell:~]$ xrgears
Error [GENERAL |  | OpenXR-Loader] : /etc/xdg/openxr/1/active_runtime.json library /nix/store/g4mhqnn6ky2s4hncf8v7z13k53kficjn-wivrn-0.11/share/openxr/1/../../../../../nix/store/g4mhqnn6ky2s4hncf8v7z13k53kficjn-wivrn-0.11/lib/libopenxr_wivrn.so does not appear to exist
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - failed to load a runtime
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : Failed to find default runtime with RuntimeInterface::LoadRuntime()
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : Failed querying extension properties
[e] xr.c:55 | Failed to enumerate instance extensions. [XR_ERROR_RUNTIME_UNAVAILABLE]
 
[e] main.cpp:395 | OpenXR graphics initialization failed.
 
```
But it would work otherwise. I'll test with the absolute paths eventually.

Reply:
`PRESSURE_VESSEL_FILESYSTEMS_RW=/etc/xdg/openxr` in the game's launch options should work. Nothing we can do about that though.

Reply:
That wouldn't work because `PRESSURE_VESSEL_FILESYSTEMS_RW=$XDG_RUNTIME_DIR/wivrn_comp_ipc` has to be defined to use WiVRn

Reply:
Hm, maybe ask Valve to fix that missing default passthrough then.

Reply:
Nevermind, that variables should theoretically be able to add a second path by separating it with a colon like so: `PRESSURE_VESSEL_FILESYSTEMS_RW="$XDG_RUNTIME_DIR/wivrn_comp_ipc:/etc/xdg/openxr/1/active_runtime.json" %command%` but I cannot get that to work. I'll try to open an issue upstream and see if I can get something done about it there